padlockcloud/sender_test.go

```diff
@@ -0,0 +1,126 @@
+		return nil
+	}
+	if err := sender.Send("test_recipient", "test_subject", "test_body"); err != nil {
+		t.Errorf("Send() == %v; want nil", err)
+	}
+}
+
+func TestEmailSenderSendFallsBackToUserWhenNoFrom(t *testing.T) {
+	sender := NewEmailSender(&EmailConfig{
+		User:     "test_user",
+		Server:   "unused_server",
+		Port:     "unused_port",
+		Password: "unused_password",
+		// No `From`: fall back to `User`.
+	})
+	sender.SendFunc = func(_ string, _ smtp.Auth, from string, _ []string, _ []byte) error {
+		if want := "test_user"; from != want {
+			t.Errorf("SendFunc from: got %q; want %q", from, want)
+		}
+		return nil
+	}
+	if err := sender.Send("test_recipient", "test_subject", "test_body"); err != nil {
+		t.Errorf("Send() == %v; want nil", err)
+	}
+
+}
+
+// This is tested separately as it relies on implementation details of smtp.PlainAuth.
+// One way to make this test less brittle is to inject the auth dependency in EmailSender.
+func TestEmailSenderSendBuildsPlainAuth(t *testing.T) {
+	sender := NewEmailSender(&EmailConfig{
+		User:     "test_user",
+		Server:   "test_server",
+		Port:     "unused_port",
+		Password: "test_password",
+	})
+	sender.SendFunc = func(addr string, a smtp.Auth, from string, to []string, msg []byte) error {
+		_, toServer, err := a.Start(&smtp.ServerInfo{
+			TLS:  true,
+			Name: "test_server",
+		})
+		if err != nil {
+			t.Fatalf("SendFunc auth: call to Start(): got %v; want nil", err)
+		}
+		if want := []byte("test_user"); !bytes.Contains(toServer, want) {
+			t.Errorf("SendFunc auth: PlainAuth toServer: got %q; want contains %q", toServer, want)
+		}
+		if want := []byte("test_password"); !bytes.Contains(toServer, want) {
+			t.Errorf("SendFunc auth: PlainAuth toServer: got %q; want contains %q", toServer, want)
+		}
+		return nil
+	}
+	if err := sender.Send("unused_recipient", "unused_subject", "unused_body"); err != nil {
+		t.Errorf("Send() == %v; want nil", err)
+	}
+}
+
+func TestEmailSenderNoUserNoAuth(t *testing.T) {
+	sender := NewEmailSender(&EmailConfig{
+		// No `User`: disable authentication.
+		Server:   "unused_server",
+		Port:     "unused_port",
+		Password: "unused_password",
+	})
+	sender.SendFunc = func(_ string, a smtp.Auth, _ string, _ []string, _ []byte) error {
+		if a != nil {
+			t.Errorf("SendFunc auth: got %q; want nil", a)
+		}
+		return nil
+	}
+	if err := sender.Send("unused_recipient", "unused_subject", "unused_body"); err != nil {
+		t.Errorf("Send() == %v; want nil", err)
+	}
+}
+
+func TestEmailSenderSendFailsWhenSendMailFails(t *testing.T) {
+	sender := NewEmailSender(&EmailConfig{
+		User:     "user",
+		Server:   "server",
+		Port:     "42",
+		Password: "password",
+		From:     "unused_from",
+	})
+	want := errors.New("jabberwock")
+	sender.SendFunc = func(string, smtp.Auth, string, []string, []byte) error {
+		return want
+	}
+	if err := sender.Send("recipient", "subject", "body"); err != want {
+		t.Errorf("sender.Send(): got %v; want %v", err, want)
+	}
+}
```

Comment:
I'm all for complete test coverage but this seems overkill for what logic where actually testing here. I do appreciate the effort though!

Reply:
Same as above. I agree this is a bit verbose but it's probably part of how I tend to write tests (it could be more compact adding some tables and indirections to test predicates and the like; but probably less readable).
I believe these are not the "bad" kind of tests where you just copy-paste the output of a real run in a test so I think they should stay but once again it's not up to me to decide.

I can agree that TestEmailSenderSendBuildsPlainAuth is probably too brittle and could go away.

Let me know whether you want me to drop everything in this file or not, and what to keep in the second case.